File: Python/video_lister/video_list.py
# SPDX-FileCopyrightText: 2022 The Video Lister Contributors

# SPDX-License-Identifier: BSD-3-Clause
"""A template to store the list of videos from a playlist into json."""

from typing import Any

from video_lister import Video


class VideoList:
    """A class to store the videos with the playlist name and id."""

    def __init__(self: "VideoList", playlist_id: str, playlist_name: str,
                 videos: list[Video]) -> None:
        self.schema: str = "https://twopizza9621536.github.io" \
                           "/schema/json/videolist.json"
        self.playlist_id: str = playlist_id
        self.playlist_name: str = playlist_name
        self.videos: list[Video] = videos

    def to_dict(self: "VideoList") -> dict[str, Any]:
        """Convert the VideoList object to a dictionary for json usage.

        Args:
            self (VideoList): The VideoList object that stores the videos in a
            schema.

        Returns:
            dict[str, Any]: The VideoList object as a dictionary.
        """
        videos_list = []

        for video in self.videos:
            videos_list.append(video.to_dict())

        return {
            "$Schema": self.schema,
            "PlaylistId": self.playlist_id,
            "PlaylistName": self.playlist_name,
            "Videos": videos_list
        }

    @staticmethod
    def to_video_list(data: dict[str, Any]) -> "VideoList":
        """Convert the formatted json data back to a VideoList object.

        Args:
            data (dict[str, Any]): The dictionary that constains the keys:
            'PlaylistId', 'PlaylistName' and 'Videos'.

        Raises:
            ValueError: If dictionary does not contain any of the following
            keys: 'PlaylistId', 'PlaylistName' or 'Videos'.

        Returns:
            VideoList: The VideoList object that was converted from a
            dictionary from json.
        """

        if ("PlaylistId" in data and "PlaylistName" in data
                and "Videos" in data):
            video_list = []
            for video in data["Videos"]:
                video_list.append(Video.to_video(video))

            return VideoList(data["PlaylistId"], data["PlaylistName"],
                             video_list)

        raise ValueError("The parsed data does not contain one or more of the "
                         "following keys:\n"
                         "'PlaylistId', 'PlaylistName' or 'Videos'")

File: dotNET/VideoLister/VideoList.cs
/*
 * SPDX-FileCopyrightText: 2021-2022 The Video Lister Contributors
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace VideoListerLibrary
{
    /// <summary>
    ///   A class to store the videos with the playlist name and id.
    /// </summary>
    public class VideoList
    {
        /// <summary>
        ///   The schema that the video list is going to be in.
        /// </summary>
        /// <value>
        ///   An URL that points to a JSON schema.
        /// </value>
        [JsonPropertyName("$Schema")]
        public string Schema { get; set; }

        /// <summary>
        ///   A reference to the play-list that the videos belongs to
        /// </summary>
        /// <value>
        ///   A long list of characters that represents the play-list in the
        ///   URL.
        /// </value>
        public string PlaylistId { get; set; }

        /// <summary>
        ///   The name of the play-list that the videos belongs to
        /// </summary>
        /// <value>
        ///   The title of the play-list.
        /// </value>
        public string PlaylistName { get; set; }

        /// <summary>
        ///   The variable to the list of videos.
        /// </summary>
        /// <value>
        ///   A IList of the <see cref="Video" />.
        /// </value>
        public IList<Video> Videos { set; get; }

        /// <summary>
        ///   A constructor to create a new instance of the VideoList object.
        /// </summary>
        /// <param name="playlistid">
        ///   The id of the play-list.
        /// </param>
        /// <param name="playlistname">
        ///   The name of the play-list.
        /// </param>
        /// <param name="videos">
        ///   A list of videos in a play-list.
        /// </param>
        public VideoList(string playlistid, string playlistname, IList<Video> videos)
        {
            PlaylistId = playlistid;
            PlaylistName = playlistname;
            Videos = videos;
        }
    }
}
